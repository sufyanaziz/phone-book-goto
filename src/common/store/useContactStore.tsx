import { createContext, useState } from "react";

type TContactStore = {
  contact: { [x: string]: unknown };
  search: string;
  offset: number;
  modal: {
    name: string;
    data: { [key: string]: any };
  };
  onSetContact: (contact: { [x: string]: unknown }) => void;
  onSetSearch: (search: string) => void;
  onSetOffset: (offset: number) => void;
  closeModal: () => void;
  openModal: (modalName: string, modalData: any) => void;
};

const ContactStore = createContext<TContactStore>({
  contact: {},
  search: "",
  offset: 0,
  modal: {
    name: "",
    data: {},
  },
  onSetContact: () => {},
  onSetSearch: () => {},
  onSetOffset: () => {},
  closeModal: () => {},
  openModal: () => {},
});

type TContactProvider = {
  children?: React.ReactNode;
};

const ContactProvider = ({ children }: TContactProvider) => {
  const [contact, setContact] = useState({});
  const [search, setSearch] = useState<string>("");
  const [offset, setOffset] = useState<number>(0);
  const [modal, setModal] = useState({ name: "", data: {} });
  return (
    <ContactStore.Provider
      value={{
        contact,
        onSetContact: setContact,
        search,
        onSetSearch: (search) => {
          setSearch(search);
          setOffset(0);
        },
        offset,
        onSetOffset: setOffset,
        modal,
        openModal: (modalName, modalData) => {
          setModal({ name: modalName, data: modalData });
        },
        closeModal: () => {
          setModal({ name: "", data: {} });
        },
      }}
    >
      {children}
    </ContactStore.Provider>
  );
};

const ContactConsumer = ContactStore.Consumer;

export { ContactStore, ContactProvider, ContactConsumer };
