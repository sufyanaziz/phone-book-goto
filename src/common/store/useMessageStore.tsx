import { TMessageId } from "@common/types/message";
import { createContext, useState } from "react";

type TMessageStore = {
  showMessage: boolean;
  activeId: TMessageId;
  setShowMessage: (activeId: TMessageId, showMessage: boolean) => void;
};

const MessageStore = createContext<TMessageStore>({
  showMessage: false,
  activeId: "",
  setShowMessage: () => {},
});

type TMessageProvider = {
  children?: React.ReactNode;
};

const MessageProvider = ({ children }: TMessageProvider) => {
  const [message, setMessage] = useState<{
    showMessage: boolean;
    activeId: TMessageId;
  }>({
    showMessage: false,
    activeId: "",
  });

  return (
    <MessageStore.Provider
      value={{
        showMessage: message.showMessage,
        activeId: message.activeId,
        setShowMessage: (activeId, showMessage) => {
          setMessage({
            activeId,
            showMessage,
          });
        },
      }}
    >
      {children}
    </MessageStore.Provider>
  );
};

const ContactConsumer = MessageStore.Consumer;

export { MessageStore, MessageProvider, ContactConsumer };
