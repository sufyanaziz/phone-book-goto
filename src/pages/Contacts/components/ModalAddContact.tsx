import React, { useContext, useEffect, useMemo, useState } from "react";
import Modal from "@common/components/Modal";
import { ContactStore } from "@common/store/useContactStore";
import { MODAL_NAME } from "../constant/modal";
import Text from "@common/components/Text";
import { css } from "@emotion/react";
import Input from "@common/components/Input";
import { AddSquareIcon, MinusSquareIcon } from "assets/icon";
import Button from "@common/components/Button";

type TModalAddContact = {
  onAddNewContact: (
    firstName: string,
    lastName: string,
    phones: string[]
  ) => void;
  loading: boolean;
  onClose: () => void;
  isComplete: boolean;
};

const ModalAddContact = ({
  onAddNewContact,
  loading,
  onClose,
  isComplete,
}: TModalAddContact) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [phoneNumbers, setPhoneNumbers] = useState<string[]>([""]);

  const { modal, closeModal } = useContext(ContactStore);

  const isDisabled = useMemo(() => {
    if (loading) return true;
    if (!firstName) return true;
    if (!lastName) return true;
    if (phoneNumbers.some((phone) => !phone)) return true;
    return false;
  }, [firstName, lastName, phoneNumbers, loading]);

  const handleCloseModal = () => {
    onClose();
    closeModal();
    setFirstName("");
    setLastName("");
    setPhoneNumbers([""]);
  };

  useEffect(() => {
    if (!loading && isComplete) {
      handleCloseModal();
    }
  }, [isComplete, loading]);

  const modalAddContactStyle = () => {
    return css({
      padding: "12px 0px",
      ".section-name": {
        ".first-name, .last-name": {
          marginBottom: "8px",
          width: "100%",
          ".input": {
            width: "100%",
          },
        },
      },
      ".header-title": {
        marginBottom: "4px",
      },
      ".section-phone-number": {
        ".phone-container": {
          display: "flex",
          marginBottom: "8px",
          width: "100%",
          "&:last-child": {
            marginBottom: "0px",
          },
          ".icons": {
            display: "flex",
            marginLeft: "8px",
            ".add-icon, .minus-icon": {
              display: "flex",
              alignItems: "center",
            },
          },
        },
      },
    });
  };

  return (
    <Modal
      isVisible={modal.name === MODAL_NAME.modalAddContact}
      header="Add New Contact"
      onClose={handleCloseModal}
      content={
        <div css={modalAddContactStyle}>
          <div className="section-name">
            <div className="first-name">
              <Text className="header-title" text="First Name" type="medium" />
              <Input
                className="input"
                placeholder="Input First Name..."
                acceptValue="string-number"
                value={firstName}
                onChange={setFirstName}
              />
            </div>
            <div className="last-name">
              <Text className="header-title" text="Last Name" type="medium" />
              <Input
                className="input"
                placeholder="Input Last Name..."
                acceptValue="string-number"
                value={lastName}
                onChange={setLastName}
              />
            </div>
          </div>
          <div className="section-phone-number">
            <Text className="header-title" text="Phones" type="medium" />
            {phoneNumbers.map((phone, listIdx) => {
              return (
                <div className="phone-container" key={`phone-${listIdx}`}>
                  <Input
                    className="input"
                    placeholder={`Phone ${listIdx + 1}`}
                    value={phone}
                    acceptValue="number-only"
                    onChange={(value) => {
                      const phones = [...phoneNumbers];
                      phones[listIdx] = value;
                      setPhoneNumbers(phones);
                    }}
                  />
                  <div className="icons">
                    {phoneNumbers.length > 1 ? (
                      <div
                        className="minus-icon"
                        onClick={() => {
                          setPhoneNumbers(
                            phoneNumbers.filter((_, idx) => idx !== listIdx)
                          );
                        }}
                      >
                        <MinusSquareIcon />
                      </div>
                    ) : null}
                    <div
                      className="add-icon"
                      onClick={() => {
                        setPhoneNumbers((prev) => [...prev, ""]);
                      }}
                    >
                      <AddSquareIcon />
                    </div>
                  </div>
                </div>
              );
            })}
          </div>
        </div>
      }
      footer={
        <div className="flex" css={{ padding: "0 12px 12px 12px" }}>
          <Button
            label="Submit"
            onClick={() => onAddNewContact(firstName, lastName, phoneNumbers)}
            disabled={isDisabled}
          />
        </div>
      }
    />
  );
};

export default ModalAddContact;
