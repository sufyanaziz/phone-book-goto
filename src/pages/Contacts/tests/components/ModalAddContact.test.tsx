import React from "react";
import renderer from "react-test-renderer";
import ModalAddContact from "@pages/Contacts/components/ModalAddContact";
import { ContactStore } from "@common/store/useContactStore";
import { mockStoreData } from "../mock/mockStore";

jest.mock("@assets/icon", () => {
  return {
    ...jest.requireActual("@assets/icon"),
  };
});

describe("Component ModalAddContact", () => {
  it("should render ModalAddContact component correctly", () => {
    const component = renderer.create(
      <ContactStore.Provider
        value={{
          ...mockStoreData,
          modal: {
            name: "modal-add-contact",
            data: {},
          },
        }}
      >
        <ModalAddContact
          isComplete
          loading
          onClose={jest.fn()}
          onAddNewContact={jest.fn()}
        />
      </ContactStore.Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
