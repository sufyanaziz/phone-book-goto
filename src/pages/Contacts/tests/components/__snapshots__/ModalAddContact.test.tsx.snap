// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ModalAddContact should render ModalAddContact component correctly 1`] = `
<div
  css={[Function]}
>
  <div
    className="modal-mask"
    onClick={[Function]}
  />
  <div
    className="modal-content"
  >
    <div
      className="header"
    >
      <div
        className=""
        css={[Function]}
      >
        Add New Contact
      </div>
      <div
        className="close-icon"
        onClick={[Function]}
      >
        <div
          className=""
          css={[Function]}
        >
          X
        </div>
      </div>
    </div>
    <div
      className="content"
    >
      <div
        css={[Function]}
      >
        <div
          className="section-name"
        >
          <div
            className="first-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              First Name
            </div>
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Input First Name..."
              value=""
            />
          </div>
          <div
            className="last-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Last Name
            </div>
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Input Last Name..."
              value=""
            />
          </div>
        </div>
        <div
          className="section-phone-number"
        >
          <div
            className="header-title"
            css={[Function]}
          >
            Phones
          </div>
          <div
            className="phone-container"
          >
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Phone 1"
              value=""
            />
            <div
              className="icons"
            >
              <div
                className="add-icon"
                onClick={[Function]}
              >
                <svg
                  fill="none"
                  height="24px"
                  viewBox="0 0 24 24"
                  width="23px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M9 12H15"
                    stroke="#435334"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                  />
                  <path
                    d="M12 9L12 15"
                    stroke="#435334"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                  />
                  <path
                    d="M3 12C3 4.5885 4.5885 3 12 3C19.4115 3 21 4.5885 21 12C21 19.4115 19.4115 21 12 21C4.5885 21 3 19.4115 3 12Z"
                    stroke="#435334"
                    strokeWidth="2"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="footer"
    >
      <div
        className="flex"
        css={
          {
            "padding": "0 12px 12px 12px",
          }
        }
      >
        <button
          className="primary"
          css={[Function]}
          disabled={true}
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
